Blockly Factory: Workspace Factory Options

evd2014 opened the pull request:
Made suggested changes to inject options in Workspace Factory, including: 

- Putting read only at the top, and hiding the trashcan if read only is selected

- Having certain zoom attributes preselected so that zoom works when first selected

- Bug fix for zoom option not working in preview

- Indent suboptions (grid, zoom, maxBlocks)

- Create a checkbox for infinite blocks, and if it's not checked, show a number input field for max number of blocks (solves problem of Infinity not being a valid in a number input).

- Manually generated string for options object to solve problem of JSON stringify putting quotations around attributes. Also indents and uses line breaks correctly. 

- Open options help tab in new window (instead of taking the user to the link in the same window and making them lose all of their work).

![screen shot 2016-08-25 at 2 26 38 pm](https://cloud.githubusercontent.com/assets/18580768/17986546/01a34084-6ad0-11e6-9499-c13fc4ffc824.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly/600)
<!-- Reviewable:end -->

evd2014 commented:
@quachtina96 @picklesrus @vicng

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/600#-:-KQ2aVHDlD3cCGlFfjQt:b-skk2sa)*
<!-- Sent from Reviewable.io -->

picklesrus commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[demos/blocklyfactory/workspacefactory/wfactory_controller.js, line 1058 \[r1\]](https://reviewable.io:443/reviews/google/blockly/600#-KQ3B7SPdlbI2gnCINJe:-KQ3B7SPdlbI2gnCINJf:b-foqxe6) ([raw file](https://github.com/google/blockly/blob/04e6f59e31f3e99864b0ff15464764122631202b/demos/blocklyfactory/workspacefactory/wfactory_controller.js#L1058)):*
> ```JavaScript
>  * @return {!Object} Blockly injection options object.
>  */
> WorkspaceFactoryController.prototype.readOptions_ = function() {
> ```

nit: you've got a bunch > 80 col now

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 148 \[r1\]](https://reviewable.io:443/reviews/google/blockly/600#-KQ38FjHNqaxF4SQrQps:-KQ38FjHNqaxF4SQrQpt:bvpwzly) ([raw file](https://github.com/google/blockly/blob/04e6f59e31f3e99864b0ff15464764122631202b/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L148)):*
> ```JavaScript
>   var addAttributes = function(obj, tabChar) {
>     var str = '';
>     var total = obj.length;
> ```

I think total is unused

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 153 \[r1\]](https://reviewable.io:443/reviews/google/blockly/600#-KQ3AiaZLJyV4_NlEK3G:-KQ39Mp1592EVm0Mm-SM:b86qqis) ([raw file](https://github.com/google/blockly/blob/04e6f59e31f3e99864b0ff15464764122631202b/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L153)):*
> ```JavaScript
>         var temp = tabChar + key + ' : {\n' + addAttributes(obj[key],
>             tabChar + '\t') + tabChar + '}, \n';
>       } else if (typeof obj[key] == 'string') {
> ```

Are there any other types that might need quotes? I'm not sure what type symbol (found in typeof documentation) is.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 165 \[r1\]](https://reviewable.io:443/reviews/google/blockly/600#-KQ38Q7xhYIcvmKZ_SNN:-KQ38Q7xhYIcvmKZ_SNO:b-b06s6e) ([raw file](https://github.com/google/blockly/blob/04e6f59e31f3e99864b0ff15464764122631202b/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L165)):*
> ```JavaScript
>   };
> 
>   return 'var options = { \n' + addAttributes(this.model.options, '\t') + '};';
> ```

can this.model.options ever be null? if so, the inner function should have a null check

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/600)*
<!-- Sent from Reviewable.io -->

evd2014 commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[demos/blocklyfactory/workspacefactory/wfactory_controller.js, line 1058 \[r1\]](https://reviewable.io:443/reviews/google/blockly/600#-KQ3B7SPdlbI2gnCINJe:-KQ71Xxoe0hZfd9s2xuq:b-f27830) ([raw file](https://github.com/google/blockly/blob/04e6f59e31f3e99864b0ff15464764122631202b/demos/blocklyfactory/workspacefactory/wfactory_controller.js#L1058)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> nit: you've got a bunch > 80 col now
</details>
Oops done.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 148 \[r1\]](https://reviewable.io:443/reviews/google/blockly/600#-KQ38FjHNqaxF4SQrQps:-KQ71Z8nMhx6xwalogmw:b-qh7p2d) ([raw file](https://github.com/google/blockly/blob/04e6f59e31f3e99864b0ff15464764122631202b/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L148)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> I think total is unused
</details>
Yep, removed. Forgot to remove that from an earlier implementation.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 153 \[r1\]](https://reviewable.io:443/reviews/google/blockly/600#-KQ3AiaZLJyV4_NlEK3G:-KQ728HjlUKgECSHEqMn:bvezaqm) ([raw file](https://github.com/google/blockly/blob/04e6f59e31f3e99864b0ff15464764122631202b/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L153)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> Are there any other types that might need quotes? I'm not sure what type symbol (found in typeof documentation) is.
</details>
There aren't any other types that need quotes. Type symbol is usually used for object attributes, and so those shouldn't need quotes for configuring the options object (and there aren't currently any in the options object, so if Blockly changes and needs values of type symbol in the attribute and they need quotes around them, then we can change this function). 

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 165 \[r1\]](https://reviewable.io:443/reviews/google/blockly/600#-KQ38Q7xhYIcvmKZ_SNN:-KQ72_adaclqilw4aKC2:b7eymvu) ([raw file](https://github.com/google/blockly/blob/04e6f59e31f3e99864b0ff15464764122631202b/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L165)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> can this.model.options ever be null? if so, the inner function should have a null check
</details>
Added an inner check just to be safe (I don't think it's possible for this to ever get called when it's null, but always safer to check).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/600)*
<!-- Sent from Reviewable.io -->
